standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/DecimalComparator.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.metastore.columnstats;
+
+import org.apache.hadoop.hive.metastore.api.Decimal;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.Comparator;
+import java.util.function.Consumer;
+
+public class DecimalComparator implements Comparator<Decimal> {
+
+  public static final DecimalComparator INSTANCE = new DecimalComparator();
+
+  public static final byte PAD_POS = (byte) 0;
+  public static final byte PAD_NEG = (byte) -1; // or 0xff in hexadecimal
+
+  public enum Approach {
+    UNKNOWN, SIGN, EQSCALE, LOG_ZERO, BITLOG_A, BITLOG_B, FALLBACK;
+
+    public static final int LEN = Approach.values().length;
+  }
+
+  /**
+   * Notifies the caller which approach was chosen.
+   *
+   * <p>It is used by tests.
+   */
+  interface ApproachInfoCallback extends Consumer<Approach> {
+  }
```

Comment:
IIUC these are only needed for testing. Is it possible to remove these by using another design:
```
interface Approach extends Comparator<Decimal> {}
class SignApproach implements Approach {}
class EqScaleApproach implements Approach {}
...
```

static Map<Predicate<Decimal>, Approach> ApproachMap;
ApproachMap.put((d1, d2) -> isNotNegative(unscaled1) != isNotNegative(unscaled2), SignApproach.INSTANCE);
...

Approach GetApproach(Decimal d1, Decimal d2) { return ApproachMap.get(); }

And unit tests could be written about the `GetApproach` method so it returns the correct Approach.
Other tests could be written for each Approach implementation.

Or we just don't test the approach individually but cover it with test cases for the `DecimalComparator.compareTo`